Sources/Core/EthereumABI/ABIElements.swift

```diff
@@ -264,72 +264,75 @@ extension ABI.Element.Function {
+    ///     - function call was aborted using `require(some_string_error_message)`: returning `["_success": false, "_abortedByRequire": true, "_errorMessageFromRequire": error_message]`.
+    ///     - in case of any error: returning `["_success": false, "_failureReason": String]`;
+    ///         Error reasons include:
+    ///            -  `outputs` declared but at least one value failed to be decoded;
+    ///            - `data.count` is less than `outputs.count * 32`;
+    ///            - `outputs` defined and `data` is empty;
+    ///            - `data` represent reverted transaction
+    public func decodeReturnData(_ data: Data) -> [String: Any] {
+        guard !outputs.isEmpty else {
+            NSLog("Function doesn't have any output types to decode given data.")
+            return ["_success": true]
+        }
 
-                return returnArray
-            }
+        /// If data is empty and outputs are expected it is treated as a `requite(expression)` call with no message.
+        /// In solidity `require(expression)` call, if `expresison` returns `false`, results in an empty response.
+        if data.count == 0 && !outputs.isEmpty {
+            return ["_success": false, "_failureReason": "Cannot decode empty data. \(outputs.count) outputs are expected: \(outputs.map { $0.type.abiRepresentation }). Was this a result of en empty `require(expression)` call?"]
         }
 
-        var returnArray = [String: Any]()
+        guard outputs.count * 32 <= data.count else {
+            return ["_success": false, "_failureReason": "Bytes count must be at least \(outputs.count * 32). Given \(data.count). Decoding will fail."]
+        }
 
-        // the "require" statement with no message argument will be caught here
-        if data.count == 0 && outputs.count == 1 {
-            let name = "0"
-            let value = outputs[0].type.emptyValue
-            returnArray[name] = value
-            if outputs[0].name != "" {
-                returnArray[outputs[0].name] = value
-            }
-        } else {
-            guard outputs.count * 32 <= data.count else { return nil }
-
-            var i = 0
-            guard let values = ABIDecoder.decode(types: outputs, data: data) else { return nil }
-            for output in outputs {
-                let name = "\(i)"
-                returnArray[name] = values[i]
-                if output.name != "" {
-                    returnArray[output.name] = values[i]
+        /// How `require(expression, string)` return value is decomposed:
+        ///  - `08C379A0` function selector for Error(string);
+        ///  - next 32 bytes are the data offset;
+        ///  - next 32 bytes are the error message length;
+        ///  - the next N bytes, where N is the int value
+        ///
+        /// Data offset must be present. Hexadecimal value of `0000...0020` is 32 in decimal. Reasoning for `BigInt(...) == 32`.
+        if data[0..<4] == Data.fromHex("08C379A0"),
```

Comment:
We should first try to decode `require` call. Any function no matter the `outputs` shall decode the message if any was given to `require`.